### PR TITLE
[GR-60173] Register custom module-info compilation participant for JDK builds without JMODs

### PR DESCRIPTION
**Pull request by jerboaa**
With [JEP 493](https://openjdk.org/jeps/493), part of JDK 24, it's possible to have JDK builds without the `jmods` folder. Currently substratevm builds assume `jmods` are always present. This patch adds an extension to mx so as to produce custom javac args when specific substratevm dependencies get compiled.

All of this is activated only when `--no-jlinking` option is being used. It's a no-op for regular GraalVM builds.

This patch depends on (which adds the needed abstractions to mx):
https://github.com/graalvm/mx/pull/287

Edit: `mx` version `7.35.1` has the needed dependency. This is what https://github.com/oracle/graal/pull/10161/commits/485f2eb9d0788d2cfe7c99be6f50d9905a996929 updates to.

Closes: https://github.com/graalvm/mandrel/issues/808

### CODE REVIEW

File: substratevm/mx.substratevm/mx_substratevm.py [hunk 1]
```diff
@@ -1193,6 +1193,19 @@ def mx_post_parse_cmd_line(opts):
...
+    # JMODs.
+    if opts.no_jlinking:
+        all_jar_dists = set()
+        for p in suite.projects_recursive():
+            if p.isJavaProject():
+                jd = p.get_declaring_module_distribution()
+                if jd:
+                    all_jar_dists.add(jd)
+        for d in all_jar_dists:
+            d.set_extra_javac_processor(NoJlinkGraalJavacArgsProcessor(d, "jdk.graal.compiler"))
```
Comment (dougxc) [inline substratevm/mx.substratevm/mx_substratevm.py:1207]: You would need this after applying https://github.com/graalvm/mx/pull/287#discussion_r1858832058:
```suggestion
            d.set_extra_javac_processor(NoJlinkGraalJavacArgsProcessor(d, "jdk.graal.compiler").__process__)
```

Comment (jerboaa) [reply substratevm/mx.substratevm/mx_substratevm.py]: Thanks

File: substratevm/mx.substratevm/mx_substratevm.py [hunk 2]
```diff
@@ -1877,6 +1890,35 @@ def checkLine(line, marker, init_kind, msg, wrongly_initialized_lines):
...
+        for m in module_desc.modulepath:
+            if m.name == self.module_name:
+                has_module_dep = True
+                graal_mod = m
+                break
+        if has_module_dep:
+            return [ '--upgrade-module-path=' + self.safe_path_arg(graal_mod.jarpath) ]
+        return []
+
+    def safe_path_arg(self, p):
```
Comment (olpaw) [inline substratevm/mx.substratevm/mx_substratevm.py:1916]: Move into `def __process__`

File: substratevm/mx.substratevm/mx_substratevm.py [hunk 3]
```diff
@@ -1877,6 +1890,35 @@ def checkLine(line, marker, init_kind, msg, wrongly_initialized_lines):
...
 
+class NoJlinkModuleInfoCompilationParticipant:
+
+    def __init__(self, dist, module_name):
+        self.dist = dist
+        self.module_name = module_name
+
+    # Upgrade module path for compilation of module-info.java files when not using jmods from the JDK
+    def __process__(self, module_desc):
+        """
```
Comment (olpaw) [inline substratevm/mx.substratevm/mx_substratevm.py:1901]: :+1: for adding this comment

**Comment by olpaw** [comment]
> @olpaw Could you help reviewing this please? Thank you!

I created an internal issue. Waiting for review comments to be addressed.

File: substratevm/mx.substratevm/mx_substratevm.py [hunk 4]
```diff
@@ -1877,6 +1890,35 @@ def checkLine(line, marker, init_kind, msg, wrongly_initialized_lines):
...
+        has_module_dep = False
+        graal_mod = None
+        for m in module_desc.modulepath:
+            if m.name == self.module_name:
+                has_module_dep = True
+                graal_mod = m
+                break
+        if has_module_dep:
+            return [ '--upgrade-module-path=' + self.safe_path_arg(graal_mod.jarpath) ]
```
Comment (olpaw) [inline substratevm/mx.substratevm/mx_substratevm.py:1913]: Once you did https://github.com/oracle/graal/pull/10161/files#r1862209294
```suggestion
            return [ '--upgrade-module-path=' + safe_path_arg(graal_mod.jarpath) ]
```

File: substratevm/mx.substratevm/mx_substratevm.py [hunk 5]
```diff
@@ -1877,6 +1890,35 @@ def checkLine(line, marker, init_kind, msg, wrongly_initialized_lines):
...
+        self.dist = dist
+        self.module_name = module_name
+
+    # Upgrade module path for compilation of module-info.java files when not using jmods from the JDK
+    def __process__(self, module_desc):
+        """
+        :param module_desc: The JavaModuleDescriptor for this distribution
+        :rtype: list of strings with extra javac arguments
+        """
+        has_module_dep = False
```
Comment (olpaw) [inline substratevm/mx.substratevm/mx_substratevm.py:1905]: Seems redundant. You can also use
```
if graal_mod:
```
below.

**Comment by jerboaa** [comment]
Thanks for the review. Working on an update.

**Comment by jerboaa** [comment]
@olpaw Updated the PR. Thanks again!

**Review by olpaw — APPROVED** [review]
Looking good. Integrating ... :hourglass_flowing_sand:

**Comment by jerboaa** [comment]
@olpaw Not sure if integration worked/works for this. Do I need to resolve the conflict or would that be counter-productive?

**Comment by olpaw** [comment]
> @olpaw Not sure if integration worked/works for this. Do I need to resolve the conflict or would that be counter-productive?

No worries. Already took care of that.

**Comment by olpaw** [comment]
Your PR is on master:
https://github.com/oracle/graal/commit/7d964fab3b3b922a2d2141f98823cb6f4b488573
Merge commit was:
https://github.com/oracle/graal/commit/36c69edf65ee12a5e43b721c7f4813419c279b63

**Comment by jerboaa** [comment]
Thank you!

**Pull request closed without merging**